sleekxmpp/plugins/xep_0313/mam.py

```diff
@@ -29,66 +154,77 @@ class XEP_0313(BasePlugin):
 
     name = 'xep_0313'
     description = 'XEP-0313: Message Archive Management'
-    dependencies = set(['xep_0030', 'xep_0050', 'xep_0059', 'xep_0297'])
+    dependencies = set(['xep_0004', 'xep_0030', 'xep_0050', 'xep_0059', 'xep_0297'])
     stanza = stanza
+        
 
     def plugin_init(self):
         register_stanza_plugin(Iq, stanza.MAM)
+        register_stanza_plugin(stanza.MAM, self.xmpp['xep_0059'].stanza.Set)
+        register_stanza_plugin(Iq, stanza.MAM_Fin)
+        register_stanza_plugin(stanza.MAM_Fin, self.xmpp['xep_0059'].stanza.Set)
+        
         register_stanza_plugin(Iq, stanza.Preferences)
         register_stanza_plugin(Message, stanza.Result)
-        register_stanza_plugin(Message, stanza.Archived, iterable=True)
         register_stanza_plugin(stanza.Result, self.xmpp['xep_0297'].stanza.Forwarded)
-        register_stanza_plugin(stanza.MAM, self.xmpp['xep_0059'].stanza.Set)
+        
 
-    def retrieve(self, jid=None, start=None, end=None, with_jid=None, ifrom=None,
-                 block=True, timeout=None, callback=None, iterator=False):
-        iq = self.xmpp.Iq()
-        query_id = iq['id']
+        self._open_queries = dict()
 
-        iq['to'] = jid
-        iq['from'] = ifrom
-        iq['type'] = 'get'
-        iq['mam']['queryid'] = query_id
-        iq['mam']['start'] = start
-        iq['mam']['end'] = end
-        iq['mam']['with'] = with_jid
 
-        collector = Collector(
-            'MAM_Results_%s' % query_id,
-            StanzaPath('message/mam_result@queryid=%s' % query_id))
-        self.xmpp.register_handler(collector)
-
-        if iterator:
-            return self.xmpp['xep_0059'].iterate(iq, 'mam', 'results')
-        elif not block and callback is not None:
-            def wrapped_cb(iq):
-                results = collector.stop()
-                if iq['type'] == 'result':
-                    iq['mam']['results'] = results
-                callback(iq)
-            return iq.send(block=block, timeout=timeout, callback=wrapped_cb)
+    def retrieve(self, start=None, end=None, with_jid=None, continue_after=None,
+                 timeout=None, callback=None, number_of_queried_elements=None, collect_all=True):
```

Comment:
Same here, arguments timeout and callback should probably stay at the end of the argument list. Also maybe add a block argument for consistency?